npm/test/upload.test.ts

```diff
@@ -0,0 +1,138 @@
+import { ClientFileProcessing } from "../src/clientprocessing"
+import { UploadFiles } from "../src/upload"
+import { GraphqlClient } from "../src/graphql"
+import { KeycloakInstance } from "keycloak-js"
+
+jest.mock("../src/clientprocessing")
+
+class ClientFileProcessingSuccess {
+  processFiles: (
+    consignmentId: number,
+    numberOfFiles: number
+  ) => Promise<number[]> = async (
+    consignmentId: number,
+    numberOfFiles: number
+  ) => {
+    const data: number[] = [1, 2]
+    return data
+  }
+  processClientFileMetadata: (
+    files: File[],
+    fileIds: number[]
+  ) => Promise<void> = async (files: File[], fileIds: number[]) => {
+    return Promise.resolve()
+  }
+}
+
+class ClientFileProcessingError {
+  processFiles: (
+    consignmentId: number,
+    numberOfFiles: number
+  ) => Promise<number[]> = async (
+    consignmentId: number,
+    numberOfFiles: number
+  ) => {
+    return Promise.reject(Error("Process files failed"))
+  }
+  processClientFileMetadata: (
+    files: File[],
+    fileIds: number[]
+  ) => Promise<void> = async (files: File[], fileIds: number[]) => {
+    return Promise.reject(Error("Process client file metadata failed"))
+  }
+}
+
+beforeEach(() => jest.resetModules())
+
+const mockKeycloak: KeycloakInstance<"native"> = {
+  init: jest.fn(),
+  login: jest.fn(),
+  logout: jest.fn(),
+  register: jest.fn(),
+  accountManagement: jest.fn(),
+  createLoginUrl: jest.fn(),
+  createLogoutUrl: jest.fn(),
+  createRegisterUrl: jest.fn(),
+  createAccountUrl: jest.fn(),
+  isTokenExpired: jest.fn(),
+  updateToken: jest.fn(),
+  clearToken: jest.fn(),
+  hasRealmRole: jest.fn(),
+  hasResourceRole: jest.fn(),
+  loadUserInfo: jest.fn(),
+  loadUserProfile: jest.fn(),
+  token: "fake-auth-token"
+}
+
+function setUpUpload(): UploadFiles {
+  const client = new GraphqlClient("test", mockKeycloak)
+  const clientFileProcessing = new ClientFileProcessing(client)
+
+  return new UploadFiles(clientFileProcessing)
+}
+
+const mockSuccess: () => void = () => {
+  const mock = ClientFileProcessing as jest.Mock
+  mock.mockImplementation(() => {
+    return new ClientFileProcessingSuccess()
+  })
+}
+
+const mockFailure: () => void = () => {
+  const mock = ClientFileProcessing as jest.Mock
+  mock.mockImplementation(() => {
+    return new ClientFileProcessingError()
+  })
+}
+
+test("retrieveConsignmentId returns consignment id from windows location", () => {
+  const url = "https://test.gov.uk/consignment/1/upload"
+  Object.defineProperty(window, "location", {
+    value: {
+      href: url,
+      pathname: "/consignment/1/upload"
+    }
+  })
+
+  const uploadFiles = setUpUpload()
+
+  expect(uploadFiles.retrieveConsignmentId()).toBe(1)
+})
+
+test("generateFileDetails adds Files to database and returns file ids", async () => {
```

Comment:
Test name nitpick: the JS only knows it's using the API to save file details. The database is an implementation detail.